### PR TITLE
[CUDACachingAllocator] Simplify update_stat and avoid casts

### PR DESCRIPTION
**Pull request by cyyever**
update_stat in CUDACachingAllocator.cpp was split into increase and decrease functions in this PR to simplify the implementation and avoid type casts throughout the code.

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -129,19 +129,25 @@ using stream_set = ska::flat_hash_set<cuda::CUDAStream>;
...
+void decrease_stat(Stat& stat, size_t amount) {
+  stat.current -= static_cast<int64_t>(amount);
   TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
       stat.current >= 0,
       "Negative tracked stat in CUDA allocator (likely logic error).");
+  stat.freed += static_cast<int64_t>(amount);
+}
 
-  stat.peak = std::max(stat.current, stat.peak);
+void update_stat(Stat& stat, int64_t amount) {
```
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:146]: Why do you preserve this one? Do we still use it?

File: c10/cuda/CUDACachingAllocator.cpp [hunk 2]
```diff
@@ -1207,28 +1213,22 @@ class DeviceCachingAllocator {
...
+            stats.inactive_split_bytes, block->size, params.stat_types);
       } else if (!block->expandable_segment_) {
         // A new split inactive block is being created from a previously unsplit
         // block, size remaining->size bytes.
         for_each_selected_stat_type(params.stat_types, [&](size_t stat_type) {
-          update_stat(
-              stats.inactive_split_bytes[stat_type],
-              static_cast<std::int64_t>(remaining->size));
-          update_stat(stats.inactive_split[stat_type], 1);
+          increase_stat(stats.inactive_split_bytes[stat_type], remaining->size);
```
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:1222]: Interesting that decrease has an _array version for this specific use case but not increase haha

File: c10/cuda/CUDACachingAllocator.cpp (hunk 1 above)
Comment (cyyever) [reply c10/cuda/CUDACachingAllocator.cpp]: Removed.

File: c10/cuda/CUDACachingAllocator.cpp (hunk 2 above)
Comment (cyyever) [reply c10/cuda/CUDACachingAllocator.cpp:1214]: Increment and decrement are not symmetrical..

**Review by albanD — COMMENTED** [review]
The change sounds good. Do we have any test about these stats collections?

**Comment by cyyever** [comment]
> The change sounds good. Do we have any test about these stats collections?

There have been already tests. In fact, some test failed when I commited an incorrect update.

**Review by albanD — APPROVED** [review]
Awesome! Thanks!

**Pull request closed without merging**